File: examples/column_generation/columngen/master.py
import gurobipy as grb

class RMP:
    def __init__(self, K, num_nodes):
        self.num_nodes = num_nodes
        self.K = K
        self.model = grb.Model("master")
        self.model.Params.LogToConsole = 0
        #self.model.Params.presolve = 0
        self.pool = list()
        return
    
    def buildModel(self, maxObj):
        model = self.model
        
        dummy_cost = maxObj
        dummy = model.addVar(lb=0.0, name="dummy", vtype=grb.GRB.CONTINUOUS)

        #Obj: minimize cost
        obj = grb.LinExpr()
        obj.add(dummy, dummy_cost) 
        model.setObjective(obj, grb.GRB.MINIMIZE)

        #Con: every node is visited at least once
        for i in range (1, self.num_nodes - 1):
            model.addLConstr(dummy >= 1, name = "cVisited[" + str(i) + "]")

        #Con: At most K vehicles are used
        model.addLConstr(dummy <= self.K, name = "cVehicles")
        
    def addColumn(self, cost, col):
        model = self.model
        self.pool.append(col)

        #Create new variable
        z = model.addVar(lb = 0.0, name = "z" + str(len(self.pool)), vtype = grb.GRB.CONTINUOUS)

        #Update objective function
        obj = model.getObjective()
        obj.add(z, cost)
        model.setObjective(obj, grb.GRB.MINIMIZE)

        #Update node constraints
        for i in range (1, len(col) - 1):
            node = col[i]
            con = model.getConstrByName("cVisited[" + str(node) + "]")
            model.chgCoeff(con, z, 1)

        #Update vehicle constraint  
        con = model.getConstrByName("cVehicles")
        model.chgCoeff(con, z, 1)

        model.update()
    
    def solve(self):
        self.model.optimize()

    def getDuals(self):
        model = self.model

        #get Mu Duals (node prize)
        mu = [0.0] * self.num_nodes
        for i in range(1, self.num_nodes - 1):
            con = model.getConstrByName("cVisited[" + str(i) +"]")
            mu[i] = con.getAttr("Pi")

        #get gamma dual
        con = model.getConstrByName("cVehicles")
        gamma = con.getAttr("Pi")

        return [mu, gamma]

    def getObj(self):
        return self.model.ObjVal

    def writeModel(self):
        self.model.write('master.mps')

File: examples/cvrp_cpp/master.cpp
#include "master.h"

RMP::RMP(int K, int num_nodes)
    : K(K), num_nodes(num_nodes), env(), model(env) {
    model.set(GRB_IntParam_LogToConsole, 0);
}

void RMP::buildModel(double maxObj) {
    GRBVar dummy = model.addVar(0.0, GRB_INFINITY, maxObj, GRB_CONTINUOUS, "dummy");

    // Every customer must be visited at least once (dummy covers infeasibility)
    for (int i = 1; i < num_nodes - 1; ++i) {
        std::string name = "cVisited[" + std::to_string(i) + "]";
        model.addConstr(dummy >= 1.0, name);
    }

    // At most K vehicles (routes) are used
    model.addConstr(GRBLinExpr(0.0) <= double(K), "cVehicles");

    model.update();
}

void RMP::addColumn(double cost, const std::vector<int>& col) {
    pool.push_back(col);

    GRBVar z = model.addVar(0.0, GRB_INFINITY, cost, GRB_CONTINUOUS, "z" + std::to_string(pool.size()));

    // Route col visits nodes col[1..size-2]; col[0] and col.back() are the depot
    for (size_t i = 1; i + 1 < col.size(); ++i) {
        int node = col[i];
        GRBConstr con = model.getConstrByName("cVisited[" + std::to_string(node) + "]");
        model.chgCoeff(con, z, 1.0);
    }

    GRBConstr con = model.getConstrByName("cVehicles");
    model.chgCoeff(con, z, 1.0);

    model.update();
}

void RMP::solve() {
    model.optimize();
}

Duals RMP::getDuals() {
    Duals duals;
    duals.mu.assign(num_nodes, 0.0);

    for (int i = 1; i < num_nodes - 1; ++i) {
        GRBConstr con = model.getConstrByName("cVisited[" + std::to_string(i) + "]");
        duals.mu[i] = con.get(GRB_DoubleAttr_Pi);
    }

    GRBConstr con = model.getConstrByName("cVehicles");
    duals.gamma = con.get(GRB_DoubleAttr_Pi);

    return duals;
}

double RMP::getObj() {
    return model.get(GRB_DoubleAttr_ObjVal);
}

void RMP::writeModel() {
    model.write("master.mps");
}
